### PR TITLE
add db4j database and kilim test

### PR DESCRIPTION
**Pull request by nqzero**
all tests green in travis (and my kilim test did run) and i've done a successful run with the 3-host configuration locally

### CODE REVIEW

**Comment by nqzero** [comment]
fix for benchmark config applied

File: toolset/databases/db4j/db4j.dockerfile [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+RUN mvn install -q
+
+FROM openjdk:11-jre-slim
+WORKDIR /db4j
+COPY --from=maven /db4j/libs libs
+COPY --from=maven /db4j/target target
+CMD java -cp target/classes:libs/\* hello.HelloData true
+
+
+
```
Comment (nbrady-techempower) [inline toolset/databases/db4j/db4j.dockerfile]: It's unlikely we're going to add this database to the toolset. https://github.com/db4j/db4j

Comment (nqzero) [reply toolset/databases/db4j/db4j.dockerfile]: thanks for taking a look at my code

what is the criteria for inclusion ? i inquired previously and was led to believe that the only requirements were that databases be client-server and persist data to disk

https://github.com/TechEmpower/FrameworkBenchmarks/issues/2209

**Comment by nbrady-techempower** [comment]
I’ll bring it up with the team but it doesn’t seem to be widely used (1 star on github?) or updated frequently and we don’t have the capacity to review that codebase without community support. Let me get back to you. 

On Monday, February 18, 2019, 8:41 AM, nqzero <notifications@github.com> wrote:


@nqzero commented on this pull request.

In toolset/databases/db4j/db4j.dockerfile:
> @@ -0,0 +1,15 @@
+FROM maven:3.5.3-jdk-11-slim as maven
+WORKDIR /db4j
+COPY pom.xml pom.xml
+RUN mvn dependency:copy-dependencies -DoutputDirectory=libs
+COPY src src
+RUN mvn install -q
+
+FROM openjdk:11-jre-slim
+WORKDIR /db4j
+COPY --from=maven /db4j/libs libs
+COPY --from=maven /db4j/target target
+CMD java -cp target/classes:libs/\* hello.HelloData true
+
+
+

thanks for taking a look at my code

what is the criteria for inclusion ? i inquired previously and was led to believe that the only requirements were that databases be client-server and persist data to disk

#2209

**Comment by nqzero** [comment]
if you can wait till friday or next week to have that discussion, i'll put together a very short pitch based on some of your old blog posts about why kilim/db4j is consistent with the type of technology the benchmark is trying to promote

**Comment by nbrady-techempower** [comment]
Sure, that seems reasonable to me. 

On Monday, February 18, 2019, 9:49 AM, nqzero <notifications@github.com> wrote:


if you can wait till friday or next week to have that discussion, i'll put together a very short pitch based on some of your old blog posts about why kilim/db4j is consistent with the type of technology the benchmark is trying to promote

**Comment by nqzero** [comment]
@nbrady-techempower here's a blurb of why i feel db4j is a good fit for the benchmark

round 17: "similar to batching but provided invisibly in the driver"
round 16: "application developers get the headroom—which is a type of luxury—to develop their application more freely"
round 15: "This flexibility is made possible when the framework and platform aren't boxing out the application" (speaking about performance, but applies more generally)

the theme that i'm focusing on is that optimizations that add complexity saddle developers and ones that don't empower developers

- in the round 17 blog you praise the Postgres query pipelining and packet inlining and opine that "more platforms will support [it]". db4j does (it's been a feature from day 1) and it's implemented without callbacks (using kilim Fibers) so it's invisible to the developer, compared to eg the vertx usage of postgres which is written in the callback passing style

- db4j queries also run inside the database as fibers, allowing them to benefit from high concurrency (several of the blog posts mention performance dropping as concurrency increases and this feature counteracts that). note: i don't know how many queries postgres will process concurrently

- the comsat (an alternative to kilim) test implementation doesn't provide routes for the database test types (evidence that there's a need for more database diversity)

- db4j moves the ORM layer into the database, and allows framework (Java) code to run in the database, in the middleware, or in the server, allowing for better data locality transparent to the application developer (analogous to the query pipelining above) without compromising the framework's expressiveness

- the db4j toolset implementation has quick startup and simple installation

- the changes required to the toolset (other than the addition of the db4j directory) are minor (get_current_world_table) and the solution is routed through the server (using a /all route). and by making this the default (as opposed to db4j-specific) it would allow adding other databases with zero changes to the python driver

- as an alternative to including db4j, would you consider allowing a "generic" database machine (eg ubuntu:16.04), that a framework could populate with a database via eg rsync and ssh

- for what it's worth, i'm a co-maintainer of kilim and the owner of the db4j portion and i'm willing to provide support. development in both is active, though neither is done primarily on github

- with the new dockerized harness (and the /all route above), the cost of an additional database doesn't seem higher than the cost of a new framework

- with regard to the requirements that were [laid out previously](https://github.com/TechEmpower/FrameworkBenchmarks/issues/2209#issuecomment-434818447), db4j runs on a separate machine and persists the data to disk

regardless of what your team decides, any feedback that you can offer in terms of which aspects were appealing or not would be useful

File: toolset/databases/db4j/src/main/java/hello/HelloCommands.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    public static class GetAll implements Taskable<HashMap<Integer,Integer>> {
+        public HashMap<Integer,Integer> task() throws Pausable {
+            return db4j().submit(txn -> {
+                HashMap<Integer,Integer> map = new HashMap();
+                worlds().getall(txn).visit(cc -> map.put(cc.key,cc.val.randomNumber));
+                return map;
+            }).await().val;
+        }
+    }
+    public static class UpdateWorlds extends GetWorlds {
```
[the remark this answers is not in the document]
Comment (nqzero) [reply toolset/databases/db4j/src/main/java/hello/HelloCommands.java]: `UpdateWorlds` extends `GetWorlds` (as in the /queries route) and calls `super.task()` which connects to the database to read the values, and then once those are ready uses a second connection to update those values. does that seem right or am i missing something ?

[the remark this answers is not in the document]
Comment (nbrady-techempower) [reply toolset/databases/db4j/src/main/java/hello/HelloCommands.java]: Nope, you're right. I somehow missed the super call.

**Comment by msmith-techempower** [comment]
I have read your proposal, but I need to digest it a bit and come to a conclusion on it. For what it's worth, all of your points make sense and seem correct to me.

**Comment by nbrady-techempower** [comment]
This db doesn't look like it's in active development. I'm going to close this for now. If there's still interest in doing this, there's a better way to add databases to the toolset now. Please check out: https://github.com/TechEmpower/FrameworkBenchmarks/wiki/Development-Adding-a-Database

**Comment by nqzero** [comment]
@nbrady-techempower Db4j is actively maintained and in use. The database abstraction looks great - exactly what I was thinking we needed. I'll refactor the pull request.

**Comment by nbrady-techempower** [comment]
@nqzero Awesome! Is there a repo other than https://github.com/db4j/db4j? I didn't see a commit since february and no releases. Is it at version 1.0 somewhere?

**Comment by nqzero** [comment]
@nbrady-techempower all dev is done in a private repo with occasional filter-syncs. that said, there are only a couple software commits since feb - most of the work has been dependencies, demos, benchmarks and legal.

i consider db4j feature complete - are there new features that you would like to see ? i'm not following semver so I'm not treating `1.0.0` as anything special (though it is a marketing opportunity that I'm hoping to take advantage of).

**Pull request closed without merging**